supporter-product-data/src/main/scala/com/gu/lambdas/ProcessSupporterRatePlanItemLambda.scala

```diff
@@ -0,0 +1,132 @@
+import com.typesafe.scalalogging.StrictLogging
+import io.circe.Codec
+import io.circe.generic.semiauto.deriveCodec
+import io.circe.parser._
+import io.circe.syntax.EncoderOps
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.{Await, Future}
+import scala.concurrent.duration.DurationInt
+import scala.util.{Failure, Success}
+
+class ProcessSupporterRatePlanItemLambda(supporterRatePlanItemProcessor: SupporterRatePlanItemProcessor)
+    extends Handler[SqsEvent, Unit] {
+
+  def this() = this(getSupporterRatePlanItemProcessor())
+
+  override protected def handlerFuture(input: SqsEvent, context: Context) = {
+    SafeLogger.info(s"Received ${input.Records.length} records from the queue")
+    Future
+      .sequence(input.Records.map { record =>
+        val subscription = decode[SupporterRatePlanItem](record.body)
+        subscription match {
+          case Right(item) => supporterRatePlanItemProcessor.processItem(item)
+          case _ =>
+            SafeLogger.warn(s"Couldn't decode a SupporterRatePlanItem with body: ${record.body}")
+            Future.successful(()) // This should never happen so I don't think it's worth alerting on
+        }
+      })
+      .map(_ => ())
+  }
+}
+
+object ProcessSupporterRatePlanItemLambda {
+  def getSupporterRatePlanItemProcessor() = {
+    val stage = StageConstructors.fromEnvironment
+    val dynamoService = SupporterDataDynamoService(stage)
+    val alarmService = AlarmService(stage)
+    val futureProcessor = ConfigService(stage).load.map { config =>
+      new SupporterRatePlanItemProcessor(
+        dynamoService,
+        alarmService,
+        new ContributionProcessor(stage, config),
+      )
+    }
+    Await.result(futureProcessor, 10.seconds)
+  }
+}
+
+class ContributionProcessor(stage: Stage, config: ZuoraQuerierConfig) {
+  val contributionIds = ContributionIds.forStage(stage)
+  lazy val zuoraService = new ZuoraSubscriptionService(config, configurableFutureRunner(60.seconds))
+
+  def isRecurringContribution(supporterRatePlanItem: SupporterRatePlanItem) =
+    contributionIds.contains(supporterRatePlanItem.productRatePlanId)
+
+  def addAmountIfContribution(supporterRatePlanItem: SupporterRatePlanItem) =
+    if (isRecurringContribution(supporterRatePlanItem)) {
+      SafeLogger.info(s"Supporter ${supporterRatePlanItem.identityId} is a recurring contributor")
+      fetchContributionAmountFromZuora(supporterRatePlanItem)
+    } else
+      Future.successful(supporterRatePlanItem)
+
+  def fetchContributionAmountFromZuora(supporterRatePlanItem: SupporterRatePlanItem) =
+    zuoraService
+      .getSubscription(supporterRatePlanItem.subscriptionName)
+      .map { sub =>
+        SafeLogger.info(
+          s"Contribution amount for supporter ${supporterRatePlanItem.identityId} is ${sub.contributionAmount}",
+        )
+        supporterRatePlanItem.copy(contributionAmount = sub.contributionAmount)
```

Comment:
`sub.contributionAmount` Option is not the same Option as `supporterRatePlanItem.contributionAmount` Option.  This can cause subtle bugs

The former means "missing charge or plan" wheras the latter means "not a contribution or supporter plus"

Personally I'd like to see us flag it in some way if we expected an amount and zuora returned missing plans etc rather than just omitting.  This isn't as bad as the one for one option collapsing but it still worries me.

Reply:
I see what you mean. I'm not sure what would be an appropriate action to take here though, I wouldn't be at all surprised to find out that we have contributions with no rate plan charges but what should we do other than just treating it as missing? We could alert but it would just get triggered every time we ran a full refresh unless we were going to go in and investigate each failing subscriptions.